lib/buffer.js

```diff
@@ -391,20 +391,53 @@ Buffer.prototype.compare = function compare(b) {
   return binding.compare(this, b);
 };
 
+function slowIndexOf(val, byteOffset, encoding) {
+  var loweredCase = false;
+  for (;;) {
+    switch (encoding) {
+      case 'utf8':
+      case 'utf-8':
+      case 'ucs2':
+      case 'ucs-2':
+      case 'utf16le':
+      case 'utf-16le':
+      case 'binary':
+        return binding.indexOfString(this, val, byteOffset, encoding);
```

Comment:
If `ParseEncoding()` is showing a significant performance hit then you could break up the calls to like `binding.indexOfUtf8String()` and on the native side have a simple wrapper that call the operation and explicitly passes the correct encoding. Look at `StringSlice()` in `src/node_buffer.cc` for an example.